### PR TITLE
Queue and store structure overhaul

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
This is a bit of a winding PR that was caught up in our OpenFF requirements and local discussion a few days ago. This PR accomplishes the following:
  - The beginnings of "locator" structures (name in progress): `{table: .., index: .., data: ...}` to give a unique way of finding data later. Implemented on the client, server, and storage socket.
 - Starts a OpenFFWorkflow class to support their requirements (will radically change).
 - Moves away from `hash_indices` as locators (in progress).
 - The queue now holds the location of the resulting data and inserted results maintain a `queue_id` to point back to the queue.
 - The above changes have ripple effects on the rest of the code.

Overall I really like the new scheme where the queue contains a complete record of what happened and these locator objects are quite handy. There are a number of issues with the current changes, but considering the depth of these changes, I would like to merge this in soon and open issues to patch this up later. 

## Todos (perhaps not this PR)
 - [ ] Implement locator returns in more places (and come up with a better name than locator).
 - [ ] Implement a locator which can search the queue and return completed results or the status of the task in question (error, queue depth, etc).
 - [ ] Switch services to use the queue id lookup rather than `hash_index` lookups (will use the above).
 - [ ] Further reduce `hash_index` to a quick search index with object comparisons on duplicate results rather than a full unique index.

## Status
- [x] Ready to go

### CODE REVIEW

File: qcfractal/interface/client.py [hunk 1]
```diff
@@ -261,7 +273,7 @@ def add_collection(self, db, overwrite=False, full_return=False):
     def get_results(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (dgasmith) [reply qcfractal/interface/client.py]: `get_results` can return multiple results. For example `get_results(program="rdkit")` would get *all* results by `rdkit`.

File: qcfractal/interface/client.py [hunk 2]
```diff
@@ -261,7 +273,7 @@ def add_collection(self, db, overwrite=False, full_return=False):
...
 
         query = {}
-        for key in ["program", "molecule_id", "driver", "method", "basis", "options"]:
+        for key in ["program", "molecule_id", "driver", "method", "basis", "options", "hash_index", "id"]:
```
[the remark this answers is not in the document]
Comment (dgasmith) [reply qcfractal/interface/client.py]: Yes, very much so. As @doaa-altarawy pointed out we are going to need to do a pretty hefty security pass.

[the remark this answers is not in the document]
Comment (Lnaden) [reply qcfractal/interface/client.py]: Okay, just checking. For now its fine, but we will make the security pass later

File: qcfractal/procedures/procedures.py [hunk 3]
```diff
@@ -86,13 +86,31 @@ def procedure_single_input_parser(storage, data):
 def procedure_single_output_parser(storage, data):
 
     # Add new runs to database
-    rdata = {k: v[0] for k, v in data.items()}
+    # Parse out hooks and data to same key/value
+    rdata = {}
+    rhooks = {}
+    for data, hooks in data:
```
[the remark this answers is not in the document]
Comment (dgasmith) [reply qcfractal/procedures/procedures.py]: No, this comes in a `[(results, hooks), (...)]` list of lists. Something to look into making consistent however.

**Comment by Lnaden** [comment]
LGTM then. Is there anything else you wanted to put in on this before we merge it?

**Comment by dgasmith** [comment]
I say merge it! I will make a few issues.

**Pull request merged**